### PR TITLE
Unblock calls on connection errors or if the stream has ended

### PR DESCRIPTION
**Pull request by prashantv**
There are 2 main fixes in this PR:
 * Make sure calls readers/writers are unblocked when a connection is closed (due to EOF or network error)
 * If an outbound stream receives EOF on the response stream, make sure the request stream gets an error

This is similar to #231 but fixes some more corner cases.

cc: @aravindvs @nomis52

### CODE REVIEW

File: inbound.go [hunk 1]
```diff
@@ -211,8 +211,15 @@ func (c *Connection) dispatchInbound(_ uint32, _ uint32, call *InboundCall, fram
...
-		if <-call.mex.ctx.Done(); call.mex.ctx.Err() == context.DeadlineExceeded {
-			call.mex.inboundTimeout()
+		select {
+		case <-call.mex.ctx.Done():
+			if call.mex.ctx.Err() == context.DeadlineExceeded {
+				call.mex.inboundTimeout()
+			}
+		case <-call.mex.errCh.c:
+			if c.log.Enabled(LogLevelDebug) {
+				c.log.Debugf("Wait for timeout/cancellation interrupted by error: %v", call.mex.errCh.err)
```
Comment (akshayjshah) [inline inbound.go]: I can't wait to get rid of this guard nonsense.

**Comment by nomis52** [comment]
I'm excited to see how this performs.

File: mex.go [hunk 2]
```diff
@@ -72,28 +132,62 @@ func (mex *messageExchange) forwardPeerFrame(frame *Frame) error {
...
-			return nil, errUnexpectedFrameType
+		if err := mex.checkFrame(frame); err != nil {
+			return nil, err
 		}
 		return frame, nil
 	case <-mex.ctx.Done():
 		return nil, GetContextError(mex.ctx.Err())
+	case <-mex.errCh.c:
+		// Select will randomly choose a case, but we want to prioritize
+		// receiving a frame over errCh. Try a non-blockign read.
```
Comment (akshayjshah) [inline mex.go]: nit: spelling for "non-blocking"

File: mex.go [hunk 3]
```diff
@@ -326,3 +447,47 @@ func (mexset *messageExchangeSet) forwardPeerFrame(frame *Frame) error {
 
 	return nil
 }
+
+// copyExchanges returns a copy of the exchanges if the exchange is active.
```
Comment (akshayjshah) [inline mex.go]: Might be helpful to note in the comment that the mexset should be locked.

File: mex.go [hunk 4]
```diff
@@ -43,28 +46,85 @@ const (
...
+	if err == nil {
+		panic("cannot Notify with no error")
+	}
+
+	// There may be some sort of race where we try to notify the mex twice.
+	select {
+	case <-e.c:
+		return fmt.Errorf("cannot broadcast error: %v, already have: %v", err, e.err)
+	default:
+	}
```
Comment (akshayjshah) [inline mex.go]: Does this guard actually protect us from races? It seems possible for concurrent callers to pass through the select statement together and reach line 72 at the same time, potentially over-writing each other.

Perhaps a double-checked lock would be more robust?

File: mex.go [hunk 5]
```diff
@@ -43,28 +46,85 @@ const (
...
 func (mex *messageExchange) forwardPeerFrame(frame *Frame) error {
+	// We want a very specific priority here:
+	// 1. Timeouts/cancellation (mex.ctx errors)
+	// 2. Whether recvCh has buffer space (non-blocking select over mex.recvCh)
+	// 3. Other mex errors (mex.errCh)
+	// Which is why we check the context error only (instead of mex.checkError).
+	// In the mex.errCh case, we do a non-blocking write to recvCh to prioritize it.
 	if err := mex.ctx.Err(); err != nil {
-		return GetContextError(err)
+		return GetContextError(mex.ctx.Err())
```
Comment (akshayjshah) [inline mex.go]: The comment above doesn't make it clear to me why we can't `return GetContextError(err)` here. Is this change required for some reason? The context docs specifically mention that successive calls to `Err()` should return the same value.

File: mex.go [hunk 6]
```diff
@@ -72,28 +132,62 @@ func (mex *messageExchange) forwardPeerFrame(frame *Frame) error {
...
 func (mex *messageExchange) recvPeerFrame() (*Frame, error) {
+	// We have to check frames/errors in a very specific order here:
+	// 1. Timeouts/cancellation (mex.ctx errors)
+	// 2. Any pending frames (non-blocking select over mex.recvCh)
+	// 3. Other mex errors (mex.errCh)
+	// Which is why we check the context error only (instead of mex.checkError)e
+	// In the mex.errCh case, we do a non-blocking read from recvCh to prioritize it.
 	if err := mex.ctx.Err(); err != nil {
-		return nil, GetContextError(err)
+		return nil, GetContextError(mex.ctx.Err())
```
Comment (akshayjshah) [inline mex.go]: Same as above - why do we need this change?

File: mex.go [hunk 7]
```diff
@@ -199,25 +315,30 @@ func (mexset *messageExchangeSet) newExchange(ctx context.Context, framePool Fra
...
+			mexset.log.WithFields(
+				LogField{"msgID", mex.msgID},
+				LogField{"msgType", mex.msgType},
+			).Warn("Attempted to create new mex after mexset shutdown")
+		} else if addErr == errDuplicateMex {
+			mexset.log.WithFields(
+				LogField{"msgID", mex.msgID},
+				LogField{"newType", mex.msgType},
+			).Warn("Duplicate msg ID for active and new mex.")
+		}
```
Comment (akshayjshah) [inline mex.go]: Should we also log any unexpected errors here?

File: mex.go [hunk 8]
```diff
@@ -175,6 +275,7 @@ type messageExchangeSet struct {
 	// maps are mutable, and are protected by the mutex.
 	exchanges        map[uint32]*messageExchange
 	timeoutExchanges map[uint32]struct{}
+	shutdown         bool
```
Comment (akshayjshah) [inline mex.go]: The placement suggests to me that we're protecting `shutdown` with the mutex too, but it doesn't look like we synchronize reads at all.

Comment (akshayjshah) [inline mex.go]: never mind

File: mex.go [hunk 9]
```diff
@@ -326,3 +447,47 @@ func (mexset *messageExchangeSet) forwardPeerFrame(frame *Frame) error {
...
+		return
+	}
+
+	for _, mex := range exchanges {
+		// When there's a connection failure, we want to notify blocked callers that the
+		// call will fail, but we don't want to shutdown the exchange as only the
+		// arg reader/writer should shutdown the exchange. Otherwise, our guarantee
+		// on sendChRefs that there's no references to sendCh is violated since
+		// readers/writers could still have a reference to sendCh even though
+		// we shutdown the exchange and called Done on sendChRefs.
```
Comment (akshayjshah) [inline mex.go]: blinds

File: mex.go (hunk 4 above)
Comment (prashantv) [inline mex.go]: It didn't protect against races very well, it was mostly to ensure we didn't call it twice. But better to do it properly, so updated to use an atomic int (it's less code, and more readable than locks IMO)

File: mex.go (hunk 5 above)
Comment (prashantv) [inline mex.go]: nope, just lots of variations of things led to me leaving this bad code in. Updating.

File: mex.go (hunk 6 above)
Comment (prashantv) [inline mex.go]: fixed, was just copy + paste from the `.Done` branch.

File: mex.go (hunk 2 above)
Comment (prashantv) [inline mex.go]: fixed typo

File: mex.go (hunk 7 above)
Comment (prashantv) [inline mex.go]: So the callers of this already log (specifically `inbound.go` and `connection.go`), these logs are here because we want to add more structured information that the caller won't know.

File: mex.go (hunk 3 above)
Comment (prashantv) [inline mex.go]: :+1:

**Comment by prashantv** [comment]
Updated

File: mex.go (hunk 7 above)
Comment (akshayjshah) [inline mex.go]: Gotcha, makes sense.

File: connection_test.go [hunk 10]
```diff
@@ -499,6 +506,38 @@ func TestWriteErrorAfterTimeout(t *testing.T) {
...
+
+		// Writes should start failing pretty soon.
+		var writeErr error
+		for i := 0; i < 100; i++ {
+			if writeErr = writeFlushStr(w, "f"); writeErr != nil {
+				break
+			}
+			time.Sleep(time.Millisecond)
+		}
+		assert.Error(t, writeErr, "Writes should fail after a connection is closed")
```
Comment (prashantv) [inline connection_test.go]: Updated to check that tchannel error code is `ErrCodeNetwork`. I don't want to get too specific on exact error message, since you could get different errors like `connection reset by peer`, `use of closed network connection` etc.

**Comment by prashantv** [comment]
Travis is failing due to a known issue (#151). Restart in Travis is not working atm, so merging without green build.
